src/Symfony/Component/DependencyInjection/Compiler/AutowiringPass.php

```diff
@@ -0,0 +1,249 @@
+                } catch (RuntimeException $e) {
+                    if (!$parameter->isDefaultValueAvailable()) {
+                        throw $e;
+                    }
+
+                    $value = $parameter->getDefaultValue();
+                }
+            }
+
+            if ($argumentExists) {
+                $definition->replaceArgument($index, $value);
+            } else {
+                $definition->addArgument($value);
+            }
+        }
+    }
+
+    /**
+     * Populates the list of available types.
+     */
+    private function populateAvailableTypes()
+    {
+        $this->types = array();
+
+        foreach ($this->container->getDefinitions() as $id => $definition) {
+            $this->populateAvailableType($id, $definition);
+        }
+    }
+
+    /**
+     * Populates the list of available types for a given definition.
+     *
+     * @param string     $id
+     * @param Definition $definition
+     */
+    private function populateAvailableType($id, Definition $definition)
+    {
+        if (!$definition->getClass()) {
+            return;
+        }
+
+        foreach ($definition->getTypes() as $type) {
+            $this->definedTypes[$type] = true;
+            $this->types[$type] = $id;
+        }
+
+        if ($reflectionClass = $this->getReflectionClass($id, $definition)) {
+            $this->extractInterfaces($id, $reflectionClass);
+            $this->extractAncestors($id, $reflectionClass);
+        }
+    }
+
+    /**
+     * Extracts the list of all interfaces implemented by a class.
+     *
+     * @param string           $id
+     * @param \ReflectionClass $reflectionClass
+     */
+    private function extractInterfaces($id, \ReflectionClass $reflectionClass)
+    {
+        foreach ($reflectionClass->getInterfaces() as $interfaceName => $reflectionInterface) {
+            $this->set($interfaceName, $id);
+
+            $this->extractInterfaces($id, $reflectionInterface);
+        }
+    }
+
+    /**
+     * Extracts all inherited types of a class.
+     *
+     * @param string           $id
+     * @param \ReflectionClass $reflectionClass
+     */
+    private function extractAncestors($id, \ReflectionClass $reflectionClass)
+    {
+        $this->set($reflectionClass->name, $id);
+
+        if ($reflectionParentClass = $reflectionClass->getParentClass()) {
+            $this->extractAncestors($id, $reflectionParentClass);
+        }
+    }
+
+    /**
+     * Associates a type and a service id if applicable.
+     *
+     * @param string $type
+     * @param string $id
+     */
+    private function set($type, $id)
+    {
+        if (isset($this->definedTypes[$type]) || isset($this->notGuessableTypes[$type])) {
+            return;
+        }
+
+        if (isset($this->types[$type])) {
+            if ($this->types[$type] === $id) {
+                return;
+            }
+
+            unset($this->types[$type]);
+            $this->notGuessableTypes[$type] = true;
```

Comment:
Just out of curiosity ; why use an associative array here, and not a simple array ? Is it because it is faster this way (I think I read something on that matter a while ago) ?